File: packages/sdk/src/messages.js
/**
 * AXIP SDK — Message Builders and Validation
 *
 * Builds AXIP protocol message envelopes for all message types.
 * Handles canonical JSON serialization for signing.
 * Validates inbound messages against the protocol spec.
 */

import { randomUUID } from 'crypto';
import { sign as signData, verify as verifyData, parsePubkey } from './crypto.js';

// ─── Known Message Types ───────────────────────────────────────

const VALID_TYPES = new Set([
  'announce', 'announce_ack',
  'discover', 'discover_result',
  'task_request', 'task_bid', 'task_accept',
  'task_result', 'task_verify', 'task_settle',
  'heartbeat', 'capability_update',
  'balance_request', 'balance_result',
  'status_request', 'status_result',
  'error'
]);

// ─── Canonical JSON for Signing ────────────────────────────────

/**
 * Produce a deterministic JSON string for signing.
 * Nulls the signature field and sorts keys at the top level.
 *
 * @param {Object} msg - AXIP message envelope
 * @returns {string} Canonical JSON string
 */
export function canonicalize(msg) {
  const signable = {
    axip: msg.axip,
    id: msg.id,
    type: msg.type,
    from: msg.from,
    to: msg.to,
    timestamp: msg.timestamp,
    nonce: msg.nonce,
    payload: msg.payload
  };
  return JSON.stringify(signable);
}

// ─── Sign and Verify ───────────────────────────────────────────

/**
 * Sign a message in place. Mutates the message to add the signature field.
 *
 * @param {Object} msg - AXIP message (signature will be set)
 * @param {Uint8Array} secretKey - Agent's ed25519 secret key
 */
export function signMessage(msg, secretKey) {
  const canonical = canonicalize(msg);
  msg.signature = signData(canonical, secretKey);
}

/**
 * Verify a message's signature against the embedded public key.
 *
 * @param {Object} msg - AXIP message with signature
 * @returns {boolean} True if signature is valid
 */
export function verifyMessage(msg) {
  if (!msg.signature || !msg.from?.pubkey) return false;
  try {
    const canonical = canonicalize(msg);
    const publicKey = parsePubkey(msg.from.pubkey);
    return verifyData(canonical, msg.signature, publicKey);
  } catch {
    return false;
  }
}

// ─── Core Message Builder ──────────────────────────────────────

/**
 * Build an unsigned AXIP message envelope.
 *
 * @param {string} type - Message type
 * @param {{ agentId: string, pubkeyFormatted: string }} from - Sender identity
 * @param {string} to - Target: agent_id, 'network', or 'relay'
 * @param {Object} payload - Type-specific data
 * @returns {Object} AXIP message envelope (unsigned)
 */
export function buildMessage(type, from, to, payload) {
  return {
    axip: '0.1.0',
    id: `msg_${randomUUID()}`,
    type,
    from: {
      agent_id: from.agentId,
      pubkey: from.pubkeyFormatted
    },
    to,
    timestamp: new Date().toISOString(),
    nonce: randomUUID(),
    payload,
    signature: null
  };
}

// ─── Type-Specific Builders ────────────────────────────────────

export function buildAnnounce(from, { capabilities, name, pricing = {}, constraints = {}, metadata = {} }) {
  return buildMessage('announce', from, 'relay', {
    capabilities,
    name,
    pricing,
    constraints,
    metadata,
    version: '0.1.0'
  });
}

export function buildDiscover(from, { capability, constraints = {} }) {
  return buildMessage('discover', from, 'relay', {
    capability,
    constraints
  });
}

export function buildDiscoverResult(from, to, { agents, requestId }) {
  return buildMessage('discover_result', from, to, {
    agents,
    request_id: requestId
  });
}

export function buildTaskRequest(from, to, { taskId, description, capability, constraints = {}, reward = 0 }) {
  return buildMessage('task_request', from, to, {
    task_id: taskId || `task_${randomUUID()}`,
    description,
    capability_required: capability,
    constraints,
    reward
  });
}

export function buildTaskBid(from, to, { taskId, bidId, price, etaSeconds, confidence, model, message }) {
  return buildMessage('task_bid', from, to, {
    task_id: taskId,
    bid_id: bidId || `bid_${randomUUID()}`,
    price_usd: price,
    estimated_time_seconds: etaSeconds,
    confidence,
    model: model || 'local',
    message: message || ''
  });
}

export function buildTaskAccept(from, to, { taskId, bidId }) {
  return buildMessage('task_accept', from, to, {
    task_id: taskId,
    bid_id: bidId
  });
}

export function buildTaskResult(from, to, { taskId, status, output, actualCost, actualTime, modelUsed }) {
  return buildMessage('task_result', from, to, {
    task_id: taskId,
    status: status || 'completed',
    output,
    actual_cost_usd: actualCost || 0,
    actual_time_seconds: actualTime || 0,
    model_used: modelUsed || 'local'
  });
}

export function buildTaskVerify(from, to, { taskId, verified, qualityScore, feedback }) {
  return buildMessage('task_verify', from, to, {
    task_id: taskId,
    verified,
    quality_score: qualityScore,
    feedback: feedback || ''
  });
}

export function buildTaskSettle(from, to, { taskId, amount, fromBalance, toBalance }) {
  return buildMessage('task_settle', from, to, {
    task_id: taskId,
    amount_usd: amount,
    settlement_method: 'credit_ledger',
    from_balance: fromBalance,
    to_balance: toBalance
  });
}

export function buildHeartbeat(from, { status = 'online', activeTasks = 0, load = 0 } = {}) {
  return buildMessage('heartbeat', from, 'relay', {
    status,
    active_tasks: activeTasks,
    load
  });
}

export function buildError(from, to, { code, message, relatedMessageId }) {
  return buildMessage('error', from, to, {
    code,
    message,
    related_message_id: relatedMessageId
  });
}

// ─── Validation ────────────────────────────────────────────────

/**
 * Validate an AXIP message envelope has required fields.
 *
 * @param {Object} msg - Message to validate
 * @returns {{ valid: boolean, error: string|null }}
 */
export function validateMessage(msg) {
  if (!msg || typeof msg !== 'object') {
    return { valid: false, error: 'Message is not an object' };
  }
  if (msg.axip !== '0.1.0') {
    return { valid: false, error: `Unknown protocol version: ${msg.axip}` };
  }
  if (!msg.id) {
    return { valid: false, error: 'Missing message id' };
  }
  if (!msg.type || !VALID_TYPES.has(msg.type)) {
    return { valid: false, error: `Invalid message type: ${msg.type}` };
  }
  if (!msg.from?.agent_id || !msg.from?.pubkey) {
    return { valid: false, error: 'Missing or incomplete from field' };
  }
  if (!msg.to) {
    return { valid: false, error: 'Missing to field' };
  }
  if (!msg.timestamp) {
    return { valid: false, error: 'Missing timestamp' };
  }
  if (!msg.nonce) {
    return { valid: false, error: 'Missing nonce' };
  }
  if (!msg.payload || typeof msg.payload !== 'object') {
    return { valid: false, error: 'Missing or invalid payload' };
  }
  return { valid: true, error: null };
}

File: packages/axip-python/src/axip/messages.py
"""
AXIP SDK — Message Builders

Builds AXIP protocol message envelopes for all message types.
Compatible with the JS SDK message format.
"""

import json
import uuid
from datetime import datetime, timezone
from typing import Any

import nacl.signing

from .crypto import Identity, sign, format_pubkey, parse_pubkey, verify, from_base64


# ─── Canonical JSON for Signing ───────────────────────────────────


def canonicalize(msg: dict) -> str:
    """
    Produce a deterministic JSON string for signing.
    Matches JS SDK canonicalize() — includes fixed top-level key order.
    """
    signable = {
        "axip": msg["axip"],
        "id": msg["id"],
        "type": msg["type"],
        "from": msg["from"],
        "to": msg["to"],
        "timestamp": msg["timestamp"],
        "nonce": msg["nonce"],
        "payload": msg["payload"],
    }
    return json.dumps(signable, separators=(",", ":"))


# ─── Sign and Verify ──────────────────────────────────────────────


def sign_message(msg: dict, signing_key: nacl.signing.SigningKey) -> None:
    """Sign a message in place — sets msg['signature']."""
    canonical = canonicalize(msg)
    msg["signature"] = sign(canonical, signing_key)


def verify_message(msg: dict) -> bool:
    """Verify a message's signature against the embedded public key."""
    if not msg.get("signature") or not msg.get("from", {}).get("pubkey"):
        return False
    try:
        canonical = canonicalize(msg)
        vk = parse_pubkey(msg["from"]["pubkey"])
        return verify(canonical, msg["signature"], vk)
    except Exception:
        return False


# ─── Core Builder ─────────────────────────────────────────────────


def build_message(
    msg_type: str,
    identity: Identity,
    to: str,
    payload: dict,
) -> dict:
    """Build an unsigned AXIP message envelope."""
    return {
        "axip": "0.1.0",
        "id": f"msg_{uuid.uuid4()}",
        "type": msg_type,
        "from": {
            "agent_id": identity.agent_id,
            "pubkey": identity.pubkey_formatted,
        },
        "to": to,
        "timestamp": datetime.now(timezone.utc).isoformat().replace("+00:00", "Z"),
        "nonce": str(uuid.uuid4()),
        "payload": payload,
        "signature": None,
    }


# ─── Type-Specific Builders ───────────────────────────────────────


def build_announce(
    identity: Identity,
    *,
    capabilities: list[str],
    name: str,
    pricing: dict | None = None,
    metadata: dict | None = None,
) -> dict:
    return build_message("announce", identity, "relay", {
        "capabilities": capabilities,
        "name": name,
        "pricing": pricing or {},
        "constraints": {},
        "metadata": metadata or {},
        "version": "0.1.0",
    })


def build_discover(
    identity: Identity,
    *,
    capability: str,
    constraints: dict | None = None,
) -> dict:
    return build_message("discover", identity, "relay", {
        "capability": capability,
        "constraints": constraints or {},
    })


def build_task_request(
    identity: Identity,
    to: str,
    *,
    task_id: str | None = None,
    description: str,
    capability: str,
    constraints: dict | None = None,
    reward: float = 0.0,
) -> dict:
    return build_message("task_request", identity, to, {
        "task_id": task_id or f"task_{uuid.uuid4()}",
        "description": description,
        "capability_required": capability,
        "constraints": constraints or {},
        "reward": reward,
    })


def build_task_bid(
    identity: Identity,
    to: str,
    *,
    task_id: str,
    bid_id: str | None = None,
    price: float,
    eta_seconds: int = 30,
    confidence: float = 0.9,
    model: str = "local",
    message: str = "",
) -> dict:
    return build_message("task_bid", identity, to, {
        "task_id": task_id,
        "bid_id": bid_id or f"bid_{uuid.uuid4()}",
        "price_usd": price,
        "estimated_time_seconds": eta_seconds,
        "confidence": confidence,
        "model": model,
        "message": message,
    })


def build_task_accept(
    identity: Identity,
    to: str,
    *,
    task_id: str,
    bid_id: str,
) -> dict:
    return build_message("task_accept", identity, to, {
        "task_id": task_id,
        "bid_id": bid_id,
    })


def build_task_result(
    identity: Identity,
    to: str,
    *,
    task_id: str,
    output: Any,
    status: str = "completed",
    actual_cost: float = 0.0,
    actual_time: int = 0,
    model_used: str = "local",
) -> dict:
    return build_message("task_result", identity, to, {
        "task_id": task_id,
        "status": status,
        "output": output,
        "actual_cost_usd": actual_cost,
        "actual_time_seconds": actual_time,
        "model_used": model_used,
    })


def build_task_verify(
    identity: Identity,
    to: str,
    *,
    task_id: str,
    verified: bool,
    quality_score: float | None = None,
    feedback: str = "",
) -> dict:
    return build_message("task_verify", identity, to, {
        "task_id": task_id,
        "verified": verified,
        "quality_score": quality_score,
        "feedback": feedback,
    })


def build_heartbeat(identity: Identity, *, status: str = "online", active_tasks: int = 0, load: float = 0.0) -> dict:
    return build_message("heartbeat", identity, "relay", {
        "status": status,
        "active_tasks": active_tasks,
        "load": load,
    })
